File: web/src/lib/permissions.ts
import type { AdminRole } from "../types";

import { roleAtLeast } from "./rbac";

export type Permission =
  | "view_monitoring"
  | "view_config"
  | "list_keys"
  | "create_key"
  | "get_key"
  | "update_key"
  | "delete_key"
  | "share_key"
  | "provisioning"
  | "key_stats"
  | "manage_users"
  | "manage_byo"
  | "list_key_requests"
  | "create_key_request"
  | "review_key_request"
  | "list_my_key_requests"
  | "update_key_description"
  | "update_key_policy"
  | "delete_own_key"
  | "delete_org_key"
  | "paste_provider_key"
  | "view_all_org_keys"
  | "create_org_key"
  | "bulk_generate_personal_keys"
  | "bulk_generate_org_keys";

const EDITOR_PLUS: Permission[] = [
  "view_monitoring",
  "view_config",
  "update_key_policy",
  "view_all_org_keys",
  "create_org_key",
];

const ADMIN_ONLY: Permission[] = [
  "paste_provider_key",
  "delete_org_key",
  "manage_users",
  "manage_byo",
  "review_key_request",
  "list_key_requests",
];

const PERSONAL_BULK: Permission[] = ["bulk_generate_personal_keys"];

const ORG_BULK: Permission[] = ["bulk_generate_org_keys"];

const VIEWER_PLUS: Permission[] = [
  "list_keys",
  "create_key",
  "get_key",
  "update_key",
  "share_key",
  "provisioning",
  "key_stats",
  "create_key_request",
  "list_my_key_requests",
  "update_key_description",
  "delete_own_key",
];

const adminOnly = new Set<Permission>(ADMIN_ONLY);
const personalBulk = new Set<Permission>(PERSONAL_BULK);
const orgBulk = new Set<Permission>(ORG_BULK);
const editorPlus = new Set<Permission>(EDITOR_PLUS);
const viewerPlus = new Set<Permission>(VIEWER_PLUS);

export function can(role: AdminRole | undefined, permission: Permission): boolean {
  if (!role) return false;
  if (adminOnly.has(permission)) {
    return role === "admin";
  }
  if (personalBulk.has(permission)) {
    return role === "viewer" || role === "admin";
  }
  if (orgBulk.has(permission)) {
    return role === "editor";
  }
  if (permission === "create_key_request") {
    return role === "viewer" || role === "editor";
  }
  if (editorPlus.has(permission)) {
    return roleAtLeast(role, "editor");
  }
  if (viewerPlus.has(permission)) {
    return roleAtLeast(role, "viewer");
  }
  return false;
}

export const permissions = {
  can,
  isViewer: (role: AdminRole | undefined) => role === "viewer",
  isEditor: (role: AdminRole | undefined) => role === "editor",
  isAdmin: (role: AdminRole | undefined) => role === "admin",
  canViewMonitoring: (role: AdminRole | undefined) => can(role, "view_monitoring"),
  canManageKeyPolicy: (role: AdminRole | undefined) => can(role, "update_key_policy"),
  canPasteProviderKey: (role: AdminRole | undefined) => can(role, "paste_provider_key"),
  canDeleteKeys: (role: AdminRole | undefined) =>
    (can(role, "delete_own_key") && role === "viewer") || can(role, "delete_org_key"),
  canDeleteOrgKeys: (role: AdminRole | undefined) => can(role, "delete_org_key"),
  canManageUsers: (role: AdminRole | undefined) => can(role, "manage_users"),
  canManageByo: (role: AdminRole | undefined) => can(role, "manage_byo"),
  canReviewKeyRequests: (role: AdminRole | undefined) => can(role, "review_key_request"),
  canRequestServiceKey: (role: AdminRole | undefined) =>
    can(role, "create_key_request"),
  requiresProvisionedKeysOnly: (role: AdminRole | undefined) => !can(role, "paste_provider_key"),
  canBulkGeneratePersonalKeys: (role: AdminRole | undefined) =>
    can(role, "bulk_generate_personal_keys"),
  canBulkGenerateOrgKeys: (role: AdminRole | undefined) =>
    can(role, "bulk_generate_org_keys"),
  canManageConfigPage: (role: AdminRole | undefined) => role === "admin",
};

export function canManageByoBans(role: AdminRole | undefined): boolean {
  return permissions.canManageByo(role);
}

export function canManageKeyPolicy(role: AdminRole | undefined): boolean {
  return permissions.canManageKeyPolicy(role);
}

File: internal/admin/permissions/permissions.go
package permissions

import (
	"strings"

	"github.com/Instawork/llm-proxy/internal/adminusers"
	"github.com/Instawork/llm-proxy/internal/apikeys"
)

// Permission names a dashboard capability. Route handlers register the minimum
// role via MinRole; business logic uses Can for finer checks inside handlers.
type Permission string

const (
	ViewMonitoring    Permission = "view_monitoring"
	ViewConfig        Permission = "view_config"
	ListKeys          Permission = "list_keys"
	CreateKey         Permission = "create_key"
	GetKey            Permission = "get_key"
	UpdateKey         Permission = "update_key"
	DeleteKey         Permission = "delete_key"
	ShareKey          Permission = "share_key"
	Provisioning      Permission = "provisioning"
	KeyStats          Permission = "key_stats"
	ManageUsers       Permission = "manage_users"
	ManageBYO         Permission = "manage_byo"
	ListKeyRequests   Permission = "list_key_requests"
	CreateKeyRequest  Permission = "create_key_request"
	ReviewKeyRequest  Permission = "review_key_request"
	ListMyKeyRequests Permission = "list_my_key_requests"

	UpdateKeyDescription     Permission = "update_key_description"
	UpdateKeyPolicy          Permission = "update_key_policy"
	DeleteOwnKey             Permission = "delete_own_key"
	DeleteOrgKey             Permission = "delete_org_key"
	PasteProviderKey         Permission = "paste_provider_key"
	ViewAllOrgKeys           Permission = "view_all_org_keys"
	CreateOrgKey             Permission = "create_org_key"
	BulkGeneratePersonalKeys Permission = "bulk_generate_personal_keys"
	BulkGenerateOrgKeys      Permission = "bulk_generate_org_keys"
)

var minRole = map[Permission]adminusers.Role{
	ViewMonitoring:    adminusers.RoleEditor,
	ViewConfig:        adminusers.RoleEditor,
	ListKeys:          adminusers.RoleViewer,
	CreateKey:         adminusers.RoleViewer,
	GetKey:            adminusers.RoleViewer,
	UpdateKey:         adminusers.RoleViewer,
	DeleteKey:         adminusers.RoleViewer,
	ShareKey:          adminusers.RoleViewer,
	Provisioning:      adminusers.RoleViewer,
	KeyStats:          adminusers.RoleViewer,
	ManageUsers:       adminusers.RoleAdmin,
	ManageBYO:         adminusers.RoleAdmin,
	ListKeyRequests:   adminusers.RoleAdmin,
	CreateKeyRequest:  adminusers.RoleViewer,
	ReviewKeyRequest:  adminusers.RoleAdmin,
	ListMyKeyRequests: adminusers.RoleViewer,
}

// MinRole returns the minimum role required to reach a route guarded by p.
func MinRole(p Permission) adminusers.Role {
	if r, ok := minRole[p]; ok {
		return r
	}
	return adminusers.RoleAdmin
}

// Can reports whether role may exercise a capability (may be stricter than MinRole).
func Can(role adminusers.Role, p Permission) bool {
	switch p {
	case ViewMonitoring, ViewConfig, UpdateKeyPolicy, ViewAllOrgKeys, CreateOrgKey:
		return role.AtLeast(adminusers.RoleEditor)
	case PasteProviderKey, DeleteOrgKey, ManageUsers, ManageBYO, ReviewKeyRequest,
		ListKeyRequests:
		return role == adminusers.RoleAdmin
	case BulkGeneratePersonalKeys:
		return role == adminusers.RoleViewer || role == adminusers.RoleAdmin
	case BulkGenerateOrgKeys:
		return role == adminusers.RoleEditor
	case CreateKeyRequest:
		return role == adminusers.RoleViewer || role == adminusers.RoleEditor
	case ListKeys, CreateKey, GetKey, UpdateKey, ShareKey, Provisioning, KeyStats,
		ListMyKeyRequests, UpdateKeyDescription, DeleteOwnKey:
		return role.AtLeast(adminusers.RoleViewer)
	default:
		return false
	}
}

// CanAccessKey reports whether the user may read or mutate a specific key record.
func CanAccessKey(role adminusers.Role, userEmail string, key *apikeys.APIKey) bool {
	if role.AtLeast(adminusers.RoleEditor) {
		return true
	}
	if key == nil || strings.TrimSpace(key.OwnerEmail) == "" {
		return false
	}
	return strings.EqualFold(key.OwnerEmail, userEmail)
}

// CanDeleteKey reports whether the user may delete the key.
func CanDeleteKey(role adminusers.Role, userEmail string, key *apikeys.APIKey) bool {
	if !CanAccessKey(role, userEmail, key) {
		return false
	}
	if Can(role, DeleteOrgKey) {
		return true
	}
	if role != adminusers.RoleViewer {
		return false
	}
	return Can(role, DeleteOwnKey) && key != nil && strings.TrimSpace(key.OwnerEmail) != ""
}

// CanCreateKeyRequest reports whether the role may submit an org key request (not admins).
func CanCreateKeyRequest(role adminusers.Role) bool {
	return Can(role, CreateKeyRequest)
}

// RequiresAutoProvision reports whether create must use auto_provision (no pasted keys).
func RequiresAutoProvision(role adminusers.Role) bool {
	return !Can(role, PasteProviderKey)
}

// UpdateKeyPolicyFields reports fields a role may set on PATCH /keys/{key}.
func UpdateKeyPolicyFieldsAllowed(role adminusers.Role) bool {
	return Can(role, UpdateKeyPolicy)
}
